c_src/sasl_auth.c

```diff
@@ -212,19 +359,31 @@ static ERL_NIF_TERM sasl_cli_step(ErlNifEnv* env, int UNUSED(argc), const ERL_NI
     if ((!enif_get_resource(env, argv[0], sasl_resource, (void**)&state))
         || (!enif_inspect_binary(env, argv[1], &challenge))) {
         return enif_make_badarg(env);
+    } else if (!sasl_auth_process_check(env, state)) {
+        return enif_raise_exception(env, ATOM_NOT_CONTROLLING_PROCESS);
     }
 
     int result = 0;
     const char* out;
     unsigned int outlen;
+    unsigned char* challenge_in = NULL;
     ERL_NIF_TERM ret;
 
     sasl_interact_t* interact = NULL;
 
-    if (state->mech_set) {
+    if (state->mech_set && state->conn) {
+        challenge_in = copy_bin(challenge);
```

Comment:
In this particular instance the debug emulator would abort warning that indeed cyrus-sasl seemed to mutate the challange, thus we avoid that possibility all together.